### ISSUE TITLE
3D: Strange shadows

### ISSUE THREAD

**Issue by benoitdm-oslandia** [qgis/QGIS#49601]
### What is the bug or the crash?

When displaying shadows with a flat or dtm terrain, a strange big black appears and no shadows are displayed inside.

Here with a dtm:
![strange_shadow_dtm](https://user-images.githubusercontent.com/64401067/183048391-4ce1c984-084a-4201-bbe4-71d7002ec16d.png)

Here without a dtm ie. all dtm layers are removed (only building layer remains), using a flat terrain:
![strange_shadow_wo_dtm](https://user-images.githubusercontent.com/64401067/183048387-235f1b53-200b-4b8d-bfa1-933528047835.png)

Also when this bug occurs, only a small scene part can have shadows:
![strange_shadow_partial_triangle](https://user-images.githubusercontent.com/64401067/183051229-14f16ef8-a88d-45b9-ac5a-cfe6a745ca64.png)

Here is a screencast:
![strange_shadow](https://user-images.githubusercontent.com/64401067/183051736-2b1323f5-7844-41ed-861b-91104a6b94ca.gif)

### Steps to reproduce the issue

Here is an example project: 
[elevation_building_issue.zip](https://github.com/qgis/QGIS/files/9267019/elevation_building_issue.zip)
[elevation_building_issue_attachments.zip](https://github.com/qgis/QGIS/files/9267020/elevation_building_issue_attachments.zip)

dtm and building are from the test data folder (tests/testdata/3d/dtm.tif and tests/testdata/3d/buildings.shp).

### Versions

Built from master source tree

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by wonder-sk** [comment]
Yeah in some cases the shadow rendering does not work well. Some of it may be bugs in the implementation, but more generally I think we need a different approach to shadow rendering (there are lots of different algorithms with different tradeoffs)
